packages/commerce-sdk-react/src/auth/index.ts

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright (c) 2022, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import {helpers, ShopperLogin, ShopperLoginTypes} from 'commerce-sdk-isomorphic'
+import jwtDecode from 'jwt-decode'
+import {ApiClientConfigParams} from '../hooks/types'
+import {BaseStorage, LocalStorage, CookieStorage} from './storage'
+
+type Helpers = typeof helpers
+interface AuthConfig {
+    shortCode: string
+    organizationId: string
+    clientId: string
+    siteId: string
+    redirectURI: string
+    proxy: string
+}
+
+interface JWTHeaders {
+    exp: number
+    iat: number
+}
+
+// this type is slightly different from ShopperLoginTypes.TokenResponse, reasons:
+// 1. TokenResponse is too generic (with & {[key:string]: any}), we need a more
+//    restrictive type to make sure type safe
+// 2. The refresh tokens are stored separately for guest and registered user. Instead
+//    of refresh_token, we have refresh_token_guest and refresh_token_registered
+type AuthDataKeys =
+    | 'access_token'
+    | 'customer_id'
+    | 'enc_user_id'
+    | 'expires_in'
+    | 'id_token'
+    | 'idp_access_token'
+    | 'refresh_token_guest'
+    | 'refresh_token_registered'
+    | 'token_type'
+    | 'usid'
+type AuthDataMap = Record<
+    AuthDataKeys,
+    {
+        storage: BaseStorage
+        key: string
+        callback?: () => void
+    }
+>
+
+/**
+ * This singleton class is used to handle shopper authentication.
+ * It is responsible for initializing shopper session, manage access
+ * and refresh tokens on server/browser environments. As well as providing
+ * a mechanism to queue network calls before having a valid access token.
+ *
+ * @Internal
+ */
+class Auth {
+    private client: ShopperLogin<ApiClientConfigParams>
+    private redirectURI: string
+    private onClient = typeof window !== 'undefined'
+    private pending: Promise<ShopperLoginTypes.TokenResponse> | undefined
+    private localStorage: LocalStorage
+    private cookieStorage: CookieStorage
+    private REFRESH_TOKEN_EXPIRATION_DAYS = 90
+
+    constructor(config: AuthConfig) {
+        this.client = new ShopperLogin({
+            proxy: config.proxy,
+            parameters: {
+                clientId: config.clientId,
+                organizationId: config.organizationId,
+                shortCode: config.shortCode,
+                siteId: config.siteId,
+            },
+        })
```

Comment:
To be consistent with the config found in the Provider, we'll need to pass in `throwOnBadResponse`:

```suggestion
        this.client = new ShopperLogin({
            proxy: config.proxy,
            parameters: {
                clientId: config.clientId,
                organizationId: config.organizationId,
                shortCode: config.shortCode,
                siteId: config.siteId
            },
            throwOnBadResponse: true
        })
```

Reply:
On 2nd thought, how about `Auth` accepting a ShopperLogin instance instead? So we can have all of the api clients initialization in one central place, which is in the Provider.

Reply:
On your 2nd thought, I'm not sure if the pro out weighs the con thou. That introduces a dependency between Auth and provider that is not necessary.